python/tests/test_stats.py

```diff
@@ -384,7 +384,123 @@ def naive_genealogical_nearest_neighbours(ts, focal, reference_sets):
+        for edge in edges_in:
+            parent[edge.child] = edge.parent
+            v = edge.parent
+            while v != -1:
+                sample_count[v] += sample_count[edge.child]
+                v = parent[v]
+
+        # Update the windows
+        assert window_index < num_windows
+        while windows[window_index] < t_right and window_index + 1 <= num_windows:
+            w_left = windows[window_index]
+            w_right = windows[window_index + 1]
+            left = max(t_left, w_left)
+            right = min(t_right, w_right)
+            span = right - left
+            # Process this tree.
+            for j, u in enumerate(focal):
+                focal_reference_set = reference_set_map[u]
+                delta = int(focal_reference_set != -1)
+                p = u
+                while p != tskit.NULL:
+                    total = np.sum(sample_count[p])
+                    if total > delta:
+                        break
+                    p = parent[p]
+                if p != tskit.NULL:
+                    scale = span / (total - delta)
+                    time_index = np.searchsorted(time_windows, time[p]) - 1
+                    if 0 <= time_index < num_time_windows:
+                        for k in range(len(reference_sets)):
+                            n = sample_count[p, k] - int(focal_reference_set == k)
+                            A[window_index, time_index, j, k] += n * scale
+                        norm[window_index, time_index, j] += span
+            assert span > 0
+            if w_right <= t_right:
+                window_index += 1
+            else:
+                # This interval crosses a tree boundary, so we update it again
+                # in the for the next tree
+                break
+
+    # Reshape norm depending on normalization selected
+    # Return NaN when normalisation value is 0
+    if span_normalise and time_normalise:
+        reshaped_norm = norm.reshape((num_windows, num_time_windows, len(focal), 1))
+    elif span_normalise and not time_normalise:
+        norm = np.sum(norm, axis=1)
+        reshaped_norm = norm.reshape((num_windows, 1, len(focal), 1))
+    elif time_normalise and not span_normalise:
+        norm = np.sum(norm, axis=0)
+        reshaped_norm = norm.reshape((1, num_time_windows, len(focal), 1))
+
+    with np.errstate(invalid="ignore", divide="ignore"):
+        A /= reshaped_norm
+    A[np.all(A == 0, axis=3)] = np.nan
+
+    # Remove dimension for windows and/or time_windows if parameter is None
+    if not windows_used and time_windows_used:
+        A = A.reshape((num_time_windows, len(focal), len(reference_sets)))
+    elif not time_windows_used and windows_used:
+        A = A.reshape((num_windows, len(focal), len(reference_sets)))
+    elif not windows_used and not time_windows_used:
+        A = A.reshape((len(focal), len(reference_sets)))
+    return A
+
+
+class TestGenealogicalNearestNeighbours(unittest.TestCase):
```

Comment:
Don't think you need the inheritance
```suggestion
class TestGenealogicalNearestNeighbours:
```